contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,419 @@
+
+  initRequest();
+
+  ProcessorState& state = encoding_state_;
+  doHeadersGo(state, headers, end_stream);
+  
+  // Headers should only happen once, so use this opportunity to add the proxy
+  // proto header, if configured.
+
+  ASSERT(route_entry_ != nullptr);
+  if (route_entry_->connectConfig().has_value()) {
+    Buffer::OwnedImpl data;
+    const auto& connect_config = route_entry_->connectConfig();
+    if (connect_config->has_proxy_protocol_config() &&
+        upstream_request_->connection().has_value()) {
+      Extensions::Common::ProxyProtocol::generateProxyProtoHeader(
+          connect_config->proxy_protocol_config(), *upstream_request_->connection(), data);
+    }
+
+    if (data.length() != 0 || end_stream) {
+      upstream_conn_data_->connection().write(data, end_stream);
+    }
+  }
+
+  // TcpUpstream::encodeHeaders is called after the UpstreamRequest is fully initialized. Alsoc use
+  // this time to synthesize the 200 response headers downstream to complete the CONNECT handshake.
+  Envoy::Http::ResponseHeaderMapPtr headersToDownstream{
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"}})};
+  upstream_request_->decodeHeaders(std::move(headersToDownstream), false);
+  return Envoy::Http::okStatus();
+}
+
+void TcpUpstream::encodeTrailers(const Envoy::Http::RequestTrailerMap&) {
+  Buffer::OwnedImpl data;
+  upstream_conn_data_->connection().write(data, true);
+}
+
+  /**
+   * Enable half-close semantics on the upstream connection. Reading a remote half-close
+   * will not fully close the connection. This is off by default.
+   */
+void TcpUpstream::enableHalfClose(bool enabled) {
+  ASSERT(upstream_conn_data_ != nullptr);
+  upstream_conn_data_->connection().enableHalfClose(enabled);
+}
+
+void TcpUpstream::readDisable(bool disable) {
+  if (upstream_conn_data_->connection().state() != Network::Connection::State::Open) {
+    return;
+  }
+  upstream_conn_data_->connection().readDisable(disable);
+}
+
+void TcpUpstream::resetStream() {
+  upstream_request_ = nullptr;
+  upstream_conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+}
+
+void TcpUpstream::onUpstreamData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream onUpstreamData data length: {}, end: {}", data.length(), end_stream);
+
+  ProcessorState& state = decoding_state_;
+  Buffer::Instance& buffer = state.doDataList.push(data);
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  GoUint64 status = dynamic_lib_->envoyGoOnUpstreamData(
+    s, end_stream ? 1 : 0, reinterpret_cast<uint64_t>(&buffer), buffer.length());
+
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(data);
+
+  if (status == static_cast<GoUint64>(UpstreamDataStatus::UpstreamDataFinish)) { // UpstreamDataFinish 
+    end_stream = true;
+    upstream_request_->decodeData(data, end_stream);
```

Comment:
Is it possible to return a different http status code downstream?
Depending on the data that received from upstream.

Reply:
yes, good idea, already support specify custom response header in envoyGoOnUpstreamData of Go side.